File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionAt.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;
import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;

public class TransformerFunctionAt extends TransformerFunction {
    public TransformerFunctionAt() {
        super(FunctionDescription.of(
            Map.of("index", ArgumentType.of(ArgType.Number).position(0))
        ));
    }

    @Override
    public Object apply(FunctionContext context) {
        var value = context.getJsonElementStreamer(null);
        var index = context.getInteger("index");
        if (index == null) {
            return null;
        }
        if (index == 0) {
            return value.stream().findFirst().orElse(null);
        }
        if (index > 0) {
            return value.stream(index.longValue(), null).findFirst().orElse(null);
        }
        // negative
        var arr = value.toJsonArray();
        var adapter = context.getAdapter();
        return adapter.get(arr, adapter.size(arr) + index);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionAt.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";

class TransformerFunctionAt extends TransformerFunction {
  constructor() {
    super({
      argsSets: [[{ name: "index", type: ArgType.Number }]],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElementStreamer(null);
    if (value == null || value.knownAsEmpty()) {
      return null;
    }
    const index = await context.getInteger("index");
    if (index == null) {
      return null;
    }
    if (index == 0) {
      return value.stream().firstOrNull();
    }
    if (index > 0) {
      return value.stream(index).firstOrNull();
    }
    // negative
    const arr = await value.toJsonArray();
    return arr.at(index) ?? null;
  }
}

export default TransformerFunctionAt;
